Refactor unit cap watch thread for default AI

relent0r opened the pull request:
## Description of the proposed changes
This pull request refactors the unitcap watch thread that runs against the AI every 30 seconds.
Original functionality.
Loop every 30 seconds and check if we are within 10 of the unit cap. If we are will kill up to 20 T1 units, on the next loop we would kill all the DIRECTFIRE point defense. Then on the next loop we would reset. This logic did not do a very good job of keeping the AI under unit cap.

fixes #6722

Refactored functionality.
Loop every 30 seconds through a data table of possible unit categories that can be culled. 
The table supports options such as 
- Perform compares before initiating the cull, for example if you want to make sure you have higher tier units before culling lower tier ones you can.
- Check if the unit is attached to a transport or  air staging facility to avoid errors
- Check if the unit is idle in case you do not want to cull units that are not idle (such as engineers that are building something)

The table can be modified and added to so that we can support better lists of units should it be desired. The table is designed to support certain mods that have units such as naval or air engineers.

The loop logic supports various configuration variables to tune its behavior. The goal is that the loop will not cull ALL of the units of the category but will dynamically decide based on how close to the unit cap we are. This will allow the loop to cull various unit categories per pass to achieve a better balance.

With the default unit cap of 1000 the AI usually does not hit the unit cap until it is on a 20km+ map size. Though there are scenarios such as water maps where the AI will build land units and not be able to transport them to the enemy.

## Testing done on the proposed changes
Have run various scenarios with various unit caps to validate that the culling logic is working correctly.

## How to test
With the default unit cap of 1000 testing can bit a little painful so I recommend using lower unit caps for testing. The idea is that if the AI can better manage its unit cap when it is lower then it will have no issues at higher unit caps. The main issues I saw when testing lower unit caps was around the issues of hitting unit cap before the AI had shifted to Tier 2. 


## Additional context
The logic is designed to work in tandem with the build condition UnitCapCheckLess. Ideally they will be tuned to work in harmony but this pull request does not retune the build condition in order to keep the code review simple. The UnitCapCheckLess will return false once the unit cap is beyond a certain amount. With the current unitCapDesiredRatio variable being set to 0.9 (90% of unit cap) the logic will not activate until the build condition is already returning false in 'most' cases.


## Checklist
- [x] Changes are annotated, including comments where useful
- [x] Changes are documented in the changelog for the next game version

relent0r committed:
refactor unit cap watch thread for default AI

relent0r committed:
update docs

relent0r committed:
remove idle check as that is 'supposed' to be handled by the getlistofunits function

relent0r committed:
add snippet

lL1l1 committed:
Annotate SetWeaponPriorites callback and usage (#6680)

Hdt80bro commented:
Are you sure it checks every 3 seconds? It only looks like 30 to me.

relent0r commented:
> Are you sure it checks every 3 seconds? It only looks like 30 to me.

Your totally right, my brain was translating it to WaitTicks despite it being quite clear :D

Hdt80bro commented on lua/AI/aiarchetype-managerloader.lua:
```diff
@@ -84,45 +84,121 @@ function SetupMainBase(aiBrain)
-    KillT1 = false
+    -- The cull table describes the types of units we want to cull if we get close to our unit cap
+    -- Options :
+    -- categories - The category of the units we want to cull
+    -- compare - If we want to validate if another type of unit already exist before we cull them
+    -- compareTo - The categories we want to compare against
+    -- checkAttached - If we want to make sure they are not attached to a transport or air staging before culling them
+    -- Remember that this table will run in order, so we want the most deisred to cull first
+    -- There is also two configurable settings
+    -- cullPressure - Indicates how many units we want to cull per pass. The closer we are to the unit cap the more units we will cull.
+    -- dynamicRatioThreshold - What sort of ratio we want when performing compares. So that we dont instantly cull lots of units just because 
```
These could be an annotation type

relent0r replied:
Have updated to be annotations on the table.

Hdt80bro commented on lua/AI/aiarchetype-managerloader.lua:
```diff
@@ -84,45 +84,121 @@ function SetupMainBase(aiBrain)
+    -- dynamicRatioThreshold - What sort of ratio we want when performing compares. So that we dont instantly cull lots of units just because 
+    -- one of the next tier is available
+
+    local cullTable = {
```
This table should be outside of the function so it can be hooked

relent0r replied:
Have externalised the table.

Hdt80bro commented on lua/AI/aiarchetype-managerloader.lua:
```diff
@@ -84,45 +84,121 @@ function SetupMainBase(aiBrain)
+        },
+        T1LandEngineer = {
+            categories = categories.MOBILE * categories.TECH1 * categories.LAND * categories.ENGINEER - categories.COMMAND,
+            compare = true,
```
The `compare` field seems unnecessary - if there's a `compareTo` field, it should be checked

relent0r replied:
Your right, I've removed it and added a compareTo check instead.

Hdt80bro commented on lua/AI/aiarchetype-managerloader.lua:
```diff
@@ -84,45 +84,121 @@ function SetupMainBase(aiBrain)
+            compare = true,
+            compareTo = categories.MOBILE * categories.LAND * categories.ENGINEER * (categories.TECH2 + categories.TECH3)- categories.COMMAND - categories.SUBCOMMANDER - categories.POD - categories.FIELDENGINEER,
+            cullRatio = 0.2,
+            checkAttached = true,
```
If `checkAttached` is always true, should it be default behavior? When would we not want to check for this?

relent0r replied:
This one, I wouldn't checkAttached if I'm working with structures or units that can't be attached. The original didn't perform this check and it wasn't until I started seeing errors when units got culled when they were on transports (or air staging) that I thought to add the check.

Hdt80bro commented on lua/AI/aiarchetype-managerloader.lua:
```diff
@@ -84,45 +84,121 @@ function SetupMainBase(aiBrain)
+        },
+    }
+    local unitCapDesiredRatio = 0.9
+    local maxCullNumber = 30
```
These would be good to leave outside the function as well

relent0r replied:
I've changed these to be parameters of the function itself since. Not sure if that offers better configuration though.

Hdt80bro commented on lua/AI/aiarchetype-managerloader.lua:
```diff
@@ -84,45 +84,121 @@ function SetupMainBase(aiBrain)
+        local cap = GetArmyUnitCap(brainIndex)
+        local capRatio = currentCount / cap
+        if capRatio > unitCapDesiredRatio then
+            local cullPressure = math.min((capRatio - 0.80) / 0.2, 1)
+            local dynamicRatioThreshold = 2.0 - (capRatio - 0.80) * 9
```
Where do the constants in this math come from? Should they be named somewhere?

relent0r replied:
I did document these in the notes of the function. I could put more details in the description if its feels too abstract to understand. These are designed to increase or limit aggressiveness of culling depending on how close you are to the limit and also to make sure the AI isn't overly aggressive at culling categories just because they are the first in the loop.

Hdt80bro commented on lua/AI/aiarchetype-managerloader.lua:
```diff
@@ -84,45 +84,121 @@ function SetupMainBase(aiBrain)
+                    if compareTo > 0 and compareFrom > 0 then
+                        local ratio = compareFrom / compareTo
+                        if ratio > dynamicRatioThreshold then
+                            local toCull = math.min(compareTo, math.ceil(compareTo * ratio * cullType.cullRatio * cullPressure))
```
Isn't `compareTo * ratio` just `compareFrom`?

relent0r commented:
@Hdt80bro thanks for the feedback, I'll go over them tonight. Some good valid points.

One question I did have was about externalizing the tables / variables.  What benefits do we gain from that? I've always been cautious about that sort of thing as its easy to 'lose' them at some point.

Hdt80bro commented:
Exposing parameters outside of functions means they aren't hidden from modders and can be easily controlled. If a mod wants to change something inside a function, they have to replace the entire thing with a copy and make their one change. This has caused problems in the past because now the mod will no longer work correctly when we update anything in that function.

There are usually also performance/memory concerns if you have to reprocess things every function call (particularly with recreating the same table structure), but that's not an issue here since the thread is only created once.

I'm not sure what you mean by things getting lost, but if you meant that it's hard to see that the function uses the variable as an upvalue, I would just put the function controls immediately above the function.

maudlin27 committed:
Include the projectile in the OnMissileIntercepted callback (#6601)

Strogoo committed:
Fix strategic bomber's 2nd formation (#6729)

relent0r commented:
> Exposing parameters outside of functions means they aren't hidden from modders and can be easily controlled. If a mod wants to change something inside a function, they have to replace the entire thing with a copy and make their one change. This has caused problems in the past because now the mod will no longer work correctly when we update anything in that function.
> 
> There are usually also performance/memory concerns if you have to reprocess things every function call (particularly with recreating the same table structure), but that's not an issue here since the thread is only created once.
> 
> I'm not sure what you mean by things getting lost, but if you meant that it's hard to see that the function uses the variable as an upvalue, I would just put the function controls immediately above the function.

Ah ok I never thought about the hooking part, makes sense thanks. I'll put some time into it this week and make some changes.

lL1l1 committed:
Fix blueprint loading crash for `nil` death animation (#6731)

relent0r committed:
URL0001 Set upgrade effects for muzzle to correct bone name (#6734)

relent0r committed:
AI - add filter for non engineering units in Engineer Manager callback (#6735)

relent0r committed:
Fix/ai incorrect function use (#6737)

Garanas committed:
Add an intermediate staging phase for deployments (#6736)

Hdt80bro committed:
Add benchmarking (#6694)

Adds in-game benchmarking to the profiler

lL1l1 committed:
Fix error when air staging gets killed when dead (#6743)

lL1l1 committed:
[Experimental] Fix factories not immediately building queue after finished upgrades or cancelled land/navy units (#6692)

Might desync, needs testing on fafdevelop.

Garanas committed:
Extract the camera screen indicator into a separate class (#6741)

This allows the indicator to be re-used for other events than just pings.

Garanas committed:
Refactor custom world rendering of shapes (#6739)

lL1l1 committed:
Fix shield assist issuing orders to an empty table (#6742)

Garanas committed:
Remove duplicated code that adjust the LODs of props during blueprint loading (#6750)

lL1l1 committed:
Add list of reviewers and snippet guidelines to the PR template (#6732)

speed2CZ committed:
Refactor lobby presets into their own file (#6687)

Yuchenjimmy committed:
Update localization tags and Traditional Chinese translations (#6618)

Update Traditional Chinese localization, add non-existing localization tags, and fix broken ones

Basilisk3 committed:
Fix various gunships and the Dragon Fly being unable to hit units that are directly beneath them (#6749)

Basilisk3 committed:
Increase the usefulness of the Cooper and compensate for previous changes to its hitbox (#6414)

Co-authored-by: FAForever <administrator@faforever.com>

relent0r committed:
update based on feedback from pull request.

Garanas committed:
Instantly pause when upgrading units by assisting them (#6691)

Garanas committed:
Add a workflow to bake blueprint properties of props (#6752)

Garanas committed:
Split up the 'Extended Graphics' game option (#6751)

Garanas committed:
Add a snippet

Garanas committed:
Revert "Add a snippet"

This reverts commit c8ba8611148c420b4a33e9832b6588f47f578141.

lL1l1 committed:
Implement decapitation victory condition (#6667)

lL1l1 committed:
Refactor the logic the share logic when an army is defeated (#5971)

This involves the logic for the full share, partial share, etc. With this rework we can better maintain it in the future.

GodFuper committed:
Disable instant shield assist for upgrading shields (#6758)

Repair orders on upgrading shields cause infinite resource drain without repairing the shield, so they should not be issued by the behavior during upgrades.

Garanas committed:
Reduce size of the cooper (#6769)

Co-authored-by: Basilisk3 <126026384+Basilisk3@users.noreply.github.com>

lL1l1 committed:
Fix minor annotation issues (#6774)

- Fixes instances where `function<>` (some undefined type) was used instead of `fun():` (a function prototype) in annotations.
- Fixes warnings of duplicate params/indices.

lL1l1 committed:
Lowercase all imports and fix import mismatch warnings (#6773)

import mismatch warnings from `Lua Diagnostics.(different-requires)`

lL1l1 committed:
Fix Aeon sniper aim while walking (#6771)

lL1l1 committed:
Balance cost and DPS stats of T3 arty (#6481)

Changes to make the DPS/mass of the T3 artillery more equal across the different factions.

Balance of arty DPS by cost:
Aeon arty set to 79k mass in proportion to old UEF arty DPS/M.
Arty costs spread evenly across 70-79k mass. Arty costs are not reduced below their current costs because that would make them come out earlier.
Damage adjusted for more per-artillery variety in shots to kill T3 shields, while also evening out STK vs different faction T3 shields.
Reload adjusted so that avg DPS vs T3 shields (this accounts for regen) per mass is equal to Aeon arty.
Energy and buildtime adjusted for the new mass costs with some rounding, but practically the same per-mass ratios.

lL1l1 committed:
Balance accuracy stats of T3 Arty (#6482)

On top of the DPS and cost issues mentioned in #6481, there are also significant issues in the effectiveness of artillery due to inaccuracy vs splash damage. T3 Artillery often comes out at the stage of the game where there are only a few highly defended key targets to attack, such as other artillery, game enders, ACUs, nukes, and SMD. This means that high splash, low accuracy, fast reload artillery is generally much less favored than the opposite which is better at breaking through shields in one point and destroying a small target. On the other hand, high splash radius with a high enough accuracy begins multiplies its damage by hitting multiple shields at the same time, which is also unfortunately countered by microing shields so only one shield is active at a time, in which case DPS becomes the most important stat (balanced in the other PR, although it assumes perfect accuracy, it is close enough).

The Firing randomness of all but the Aeon T3 artillery is decreased to make them more accurate. The Cybran RateOfFire is adjusted from its initial changes in the DPS/mass cost changes of #6481
UEF Duke:
    - Firing Randomness: 0.525 -> 0.467

Seraphim Hovatham:
    - Firing Randomness: 0.675 -> 0.560

Cybran Disruptor:
    - Firing Randomness: 0.75 -> 0.646
    - RateOfFire: 10/72 -> 10/75 (10/integer interval in ticks)

lL1l1 committed:
Adjust Othuum hitbox and DPS distribution (#6728)

Allocate some DPS from the Othuums short range weapons to its longer range weapon and reduce the muzzle velocity of the long range weapon. More DPS on its longer range weapon should help it with trading better when it has difficulty catching up to the enemy units such as when fighting Harbingers. In order to improve the Othuums pathing its hitbox is also made a bit shorter to cause less bumping between the Othuums when moving.

Othuum: T3 Siege Tank (XSL0303):
Thau Cannon:
      - Reload time: 4.0 -> 3.4 seconds (DPS 156 -> 184)
      - Muzzle velocity: 40 -> 36
Aire-au Bolters: (x2)
      - Damage: 64 -> 57 (DPS 256 -> 228)

Garanas committed:
Fix icons in the map preview of the autolobby being misplaced when using UI scaling (#6768)

Basilisk3 committed:
Preliminary tweaks for TML and TMD (#6738)

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

lL1l1 committed:
Adjust aurora handling (#6486)

lL1l1 committed:
Revert "Adjust aurora handling" (#6782)

Basilisk3 committed:
Fix various anti-air weapons being unable to engage units that fly right above them (#6753)

Tagada14 committed:
Fix Solace torpedo damage and re-implement its splitting behavior (#6522)

- Now correctly splits into 2 projectiles instead of 3 (dmg 6k -> 4k)
- Uses the SplitComponent
- Parent projectiles deal full damage now if they fall directly onto a target
Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

Basilisk3 committed:
Nerf short-range teleport jumps (#6623)

lL1l1 committed:
Increase Paragon production update rate (#6671)

K-ETFreeman committed:
Revert increased size of Seraphim air transport footprints (#6755)

This re-introduces an issue with them not moving off building locations, but fixes a more important issue of them not wanting to land on certain plateaus.
An ideal fix would move the model around so that it can have a footprint more centered on its hitbox/where units are attached.

Garanas committed:
Refactor victory condition to reduce complexity (#6766)

Applies the singleton and factory pattern to make it easier to reason about.

Garanas committed:
Add ability to paint directly on the map and share that with allies (#6726)

lL1l1 committed:
Increase Solace damage per volley from 4k to 5k (#6785)

When balance team discussed the Solace we generally agreed that at 4k damage it was too weak, but too strong with the bug that made it deal 6k damage, so we want to try out 5k damage.

Solace: T3 Torpedo Bomber (XAA0306)

    Torpedo Cluster
        Damage per volley: 4000 -> 5000

lL1l1 committed:
Reduce omni radius of Aeon ACU Sensors (#6745)

Make fire beetles more viable against Aeon by reducing the omni range of the sensors upgrade so that the ACU has to stay at the front to use the omni effectively, which allows counterplay since the ACU and army are possible to threaten. The range is enough to match gun advanced range.

Aeon ACU (UAL0001) Sensor System:

    Omni radius: 80 -> 36

lL1l1 committed:
Buff Cybran navy's missile deflectors (#6744)

Cybran navy is unable to properly defend itself against missile ships after they were changed to deflector type TMD with heavily nerfed fire rate and range.

Increase the fire rate and range to match standard naval TMD.

Cybran cruiser/carrier TMD:

    Range: 20/26 -> 44
    Max target height: 8/10 -> infinite
    Reload time: 4/2.5 seconds -> 1.9 seconds

lL1l1 committed:
Buff UEF Destroyer DPS/speed/turn rate at cost of HP (#6690)

With the current UEF destroyer rebalance, it still has issues in navy because while it does have superior HP and gun DPS, it does not have the speed to force favorable fights, its short range loses it a lot of HP over time, and the gun DPS advantage is equalized once torpedo DPS comes into play.

UEF Destroyer (UES0201):

    Help the destroyer take better fights:
        Max speed/acceleration/braking: 5.0 -> 5.5
        Turn rate: 50 -> 55
    Buff DPS and revert the HP buff so that more proactive usage of the destroyer is better rewarded:
        Gauss cannon damage (DPS): 275 (275) -> 305 (305)
            It can now 6 shot exodus (6600 -> 7320 damage per 6 volleys); it is unlikely due to inaccuracy, but it is worth noting.
        HP: 8000 -> 7200

lL1l1 committed:
Prevent weapons that fire after charging from charging while not aimed (#6791)

lL1l1 committed:
Fix cancelling tempest firing cycle (#6790)

lL1l1 committed:
Fix errors when self destruct is called with 0 units (#6788)

lL1l1 committed:
Change Aeon ACU's personal shield from blue to green (#6794)

Co-authored-by: Isakentyi <kfat2@mail.ru>

BlackYps committed:
Fix TTerrainXPExt (#6770)

Garanas committed:
Fix a bug where recall does not work (#6802)

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

Yuchenjimmy committed:
Fix typo's in Chinese translations (#6801)

BlackYps committed:
Prepare release 3822 (#6800)

Co-authored-by: Rhaelya <rhaelya@proton.me>
Co-authored-by: Basilisk3 <126026384+Basilisk3@users.noreply.github.com>
Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

lL1l1 committed:
Fix matchmaker games from the autolobby not sharing units on death (#6810)

lL1l1 committed:
Fix error when unit with no fire rate is shared (#6807)

lL1l1 committed:
Fix issues with transport speed reduction (#6814)

Adds a minimum speed
Fixes air units and structures attached to transports causing errors when the next unit is attached

lL1l1 committed:
Fix WrapText not picking up trailing single characters after newlines (#6816)

lL1l1 committed:
Fix localization for disconnect delay being used by disconnect share options (#6809)

Garanas committed:
Re-implement announcements (#6813)

Introduces a new setup that makes it easier to create new type of announcements.

Garanas committed:
Revert delay to pause to 5 ticks when assisting to upgrade (#6819)

lL1l1 committed:
Implement proper transfer of unbuilt units in factories/external factories (#6784)

Garanas committed:
Prepare for game version 3823 (#6820)

Garanas committed:
Fix the reference branch when deploying the FAF gametype

It was referencing the old 'master' branch, which we renamed to
'staging/faf'

lL1l1 committed:
Fix factories destroying finished built units (#6824)

Garanas committed:
Changelog for game version 3824 (#6825)

4z0t committed:
Proper type annotation for `GetCommandMode` (#6823)

Basilisk3 committed:
Update the description of the Aeon ACU's vision upgrade (#6828)

lL1l1 committed:
Fix desyncs in replays and/or multiplayer when using debug functionality (#6830)

ostrovaya committed:
Reduce memory allocations when unit formations are computed (#6826)

lL1l1 committed:
Fix a possible cause of desyncs (#6835)

lL1l1 committed:
Sanitize input for some lobby WARN messages (#6796)

lL1l1 committed:
Fix errors when done building the Eye of Rhianne while stalling energy (#6844)

lL1l1 committed:
Annotate callbacks related to the use of the escape hotkey (#6799)

lL1l1 committed:
Move teleport reticle out of command mode data (#6797)

lL1l1 committed:
Improve annotations related to Buffs (#6837)

lL1l1 committed:
Annotation fixes for the profiler (#6832)

lL1l1 committed:
Annotate worldview control and ping data (#6798)

Garanas committed:
Add a game option to tweak the rendering distance for shadows (#6763)

Garanas committed:
Fix various issues with announcements (#6845)

ostrovaya committed:
Reduce conditional table accesses of projectiles when colliding or losing targets (#6853)

Co-authored-by: ostrovaya <sondery@pm.me>
Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

sting-2 committed:
Map Textures Wasteland Biome - Reduced Size (#6803)

Basilisk3 committed:
Adjust size of Bulwark and the Mermaid back towards their Steam values (#6870)

ostrovaya committed:
Remove unused field from projectile trajectory correction functionality (#6864)

ostrovaya committed:
Reduce table access count in `OnImpact` of projectiles (#6862)

Garanas committed:
Extend the rendering distance of all tree (groups) (#6851)

These changes are only possible because shadows no longer render indefinitely.

Basilisk3 committed:
Increase the `StrategicIconSize` of the Seraphim ACU's OC projectile to match that of the other ACUs (#6871)

relent0r committed:
Fix : AI - Correct SurfaceThreatLevel values for all ACU units (#6868)

ostrovaya committed:
Reuse tables when creating beam weapons instead of allocating new ones (#6866)

Co-authored-by: ostrovaya <sondery@pm.me>

lL1l1 committed:
Adjust Aeon range/speed gun upgrade costs (#6705)

Garanas committed:
Fix issues with victory results (#6840)

lL1l1 committed:
Fix continental under construction blocking pgen construction (#6884)

lL1l1 committed:
Fix Cybran engis not turning wheels back to land mode (#6879)

4z0t committed:
Remove redundant t1 Hive category (#6891)

Co-authored-by: Basilisk3 <126026384+Basilisk3@users.noreply.github.com>

ostrovaya committed:
Upvalue and localize frequently used functions and tables in formations (#6854)

Co-authored-by: ostrovaya <sondery@pm.me>

ostrovaya committed:
Reduce weapon damage table computations when applying buffs by storing constant data (#6860)

Co-authored-by: ostrovaya <sondery@pm.me>
Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

BlackYps committed:
Update deployment instructions (#6580)

Co-authored-by: (Jip) Willem Wijnia <applicaties@jipwijnia.nl>
Co-authored-by: (Jip) Willem Wijnia <w.b.wijnia@gmail.com>

lL1l1 committed:
Fix recall desync when an army is defeated the same tick a recall vote starts (#6898)

lL1l1 committed:
Reduce Cybran T1 mobile AA's damage (#6843)

The damage of the Cybran T1 Mobile Anti-Air Gun is lowered since the projectiles can retarget now. The AA doesn't need so much DPS to stay equal with other factions anymore.
DPS goes from 32 to 28. UEF/Sera AA have 26 DPS and Aeon has 24 DPS with 3 more range.

    Sky Slammer: T1 Mobile Anti-Air Gun (URL0104)
        Nanodart Launcher (Anti Air)
            Damage: 16 -> 14

lL1l1 committed:
Increase Seraphim T1 sub torpdef reload time (#6880)

Sera T1 sub torpdef is currently too strong. It easily beats destroyers mass for mass, dominates other T1 subs, and can even take on stationary torpedo defenses.

13s reload time is chosen since Sera subs get even better at higher counts, the sera player is more likely to spam tons of them compared to other factions spamming subs, and you can reload the antitorp out of combat for free HP saved.

An even slower reload is not chosen despite beating destroyers mass for mass because Cybran destros can be supported by T1/T2 subs to deplete antitorps, and other factions have their own ways to beat Sera T1 subs at T2 (UEF cooper, Aeon depth charges/T2 sub, Sera subs/extra destroyer torps). Curbing the snowball at T1 should allow a transition into T2 to be possible, even if the Sera subs can find an advantage eventually at T1.

These changes should make stacking a lot of Seraphim T1 submarines in a single spot weaker, allowing torpedos to deal more damage to them.

Seraphim T1 Submarine (XSS0203):

    Torpedo defense reload: 5 sec -> 13 sec

Basilisk3 committed:
Increase the `MuzzleSalvoSize` of the Cybran ACU's Nanite Torpedo upgrade to 4 (#6895)

Further increase the MuzzleSalvoSize of the Cybran ACU's Nanite Torpedo upgrade to 4, as it still had difficulties penetrating torpedo defenses after (#6476) increased it to 3. It's DPS remains unchanged.

Cybran ACU (URL0001)

    Nanite Torpedo Launcher
        Damage (DPS): 60 (225) --> 45 (225)
        MuzzleSalvoSize: 3 --> 4

Basilisk3 committed:
Reduce the Additional Regen Rate the Cybran ACU's Nano-Repair System provides (#6894)

RabidPope committed:
Add ability for Soul Ripper weapons to fire at gunships (#6435)

4z0t committed:
Check whether LazyVar.OnDirty exists in case it was destroyed earlier (#6885)

lL1l1 committed:
Improve single player command line launch (#6873)

lL1l1 committed:
Integrate Saver's Aeon T2 -> T3 Shield upgrade mod (#6877)

lL1l1 committed:
Make structure hp decay quickly below 10% completion (#6887)

Blocking with PD is considered unfun and has a bad aesthetic. This is because it can allow for cheap ACU kills by getting enemy ACUs stuck in rapidly formed walls of unbuilt PD, which decay slowly unlike low cost walls (unit build progress decays at 0.1 resource per tick from the highest out of the mass/energy/buildtime requirements).

Implement a component that lets a unit start decaying quickly while it is below a given fraction complete.
    - The default decay rate is equivalent to how long walls would take to decay if built with the same buildpower.
    - The default completion threshold is equivalent to how long walls take to build if built with the same buildpower.

BlackYps committed:
water absorption improvements (#6858)

BlackYps committed:
Prepare release 3825 (#6899)

Co-authored-by: Basilisk3 <126026384+Basilisk3@users.noreply.github.com>
Co-authored-by: Rhaelya <rhaelya@proton.me>

BlackYps committed:
Improve documentation (#6904)

Garanas committed:
Fix an edge case where announcements would break cinematics (#6905)

BlackYps committed:
Prepare hotfix

4z0t committed:
Remove access to a potentially dangerous engine function (#6909)

Garanas committed:
Removes some lingering print statements (#6911)

Garanas committed:
Improve annotations of the trail blueprint (#6910)

Isocinetic committed:
Fix typo in UEF T2 transport RU description (#6902)

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

4z0t committed:
Refactor SCX menual panel border creation into UIUtil function (#6857)

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

Hdt80bro committed:
Fix `RemoveOutliers` statistics (#6848)

lL1l1 committed:
Update Lua Library function annotations (#6908)

Basilisk3 committed:
Remove the Seraphim Lightning Tank's `FiringTolerance` (#6900)

Garanas committed:
Prepare release for game version 3827 (#6914)

lL1l1 committed:
Fix Salem's overly long antitorp target check interval (#6921)

Salem wasn't blocking torpedo bombers correctly. This was caused by Salem having a set 1.0 target check interval for its antitorpedo flares, which is much higher than the default 0.4s interval for anti-projectile weapons.

BlackYps committed:
Add high resolution resource markers (#6918)

lL1l1 committed:
Fix unit view detail error when nuke damage isn't written properly in blueprint (#6865)

lL1l1 committed:
Allow queuing Aeon shield upgrades before it is built (#6916)

lL1l1 committed:
Fix unkillable recalling ACU when a player disconnects in the middle of recall (#6919)

Garanas committed:
Log all received message types in the lobby (#6923)

lL1l1 committed:
Automatically rebuild novax satellite when destroyed for human armies (#6920)

Garanas committed:
Prepare release for game version 3827 (#6925)

Garanas committed:
Fix the date of the changelog of game version 3828

Garanas committed:
Add debug capabilities for ICE/WebRTC on non-production branches

Garanas committed:
Adjust default maximum frame time to not interfere with vsync (#6932)

Garanas committed:
Add lobby option to toggle overflow (#6938)

lL1l1 committed:
Update blueprint modding documentation (#6883)

lL1l1 committed:
Fix typo in #6938's snippet's filename (#6940)

maudlin27 committed:
Fix torrent attackmove range (#6712)

Garanas committed:
Add annotations about the sensitive nature of the guard scan radius unit blueprint value (#6943)

lL1l1 committed:
Annotate `OnMouseButtonPress` and related code (#6942)

dependabot[bot] committed:
Bump rexml from 3.3.9 to 3.4.2 in /docs (#6927)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Basilisk3 committed:
Increase the `StrategicIconSize` of the Obsidian's Quantum Cannon and the Wagner's Meson Rocket (#6946)

BlackYps committed:
Terrain301 compatibility (#6945)

4z0t committed:
Improve resolution of available support factories of ACUs upgrading to higher tech suites (#6892)

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

speed2CZ committed:
Fix replay ID extraction possibly breaking UI (#6947)

lL1l1 committed:
Adjust T2 HQ HP so that Cybran HQs do not die to 1 TML hit (#6936)

Adjust HP of T2 Air/Land HQs so that Cybran T2 HQs do not die to 1 TML hit. This prevents Cybran players from being unfairly locked out of the tech needed to counter TMLs.

T2 HQ HPs: 5.5k, 6.4k, 7k, 8k -> 6.1k, 6.6k, 7.2k, 8k
Cybran T2 HQ Regen rate: 20 -> 15
Regen is reduced because the HP difference between Cybran and other factions is reduced.

lL1l1 committed:
ACU TML Minimum range 15 -> 30 (#6935)

Increase minimum range of ACU TMLs so that they are less powerful in ACU vs ACU fights.

UEF and Seraphim ACU TML:
    Minimum range: 15 -> 30

UEF ACU Billy nuke:
    Minimum range: 15 -> 30
    The range has to be increased so that the minimum range ring for the TML shows up.

lL1l1 committed:
ACU TML Missile BT 450 -> 700 (#6934)

Increase build time for ACU TMLs so that they are less effective on the move, in combat, and from the water. You can still reach a high fire rate with engineer assistance.

    Seraphim and UEF ACU TML:
        Missile build time: 450 -> 700 (same as SACU)
            Build time with T2 suite: 10.8 -> 16.7 seconds
            Build time with T3 suite: 4.5 -> 7.0 seconds

Isocinetic committed:
Add energy consumption for beetles cloak (#6897)

Adds energy upkeep to the beetles cloaking ability as they were the only cloaked unit without upkeep costs. Also adds the ability to turn off the cloaking to save the energy cost.
This should weaken making but not using larger amounts of beetles while keeping them cloaked, especially after a T2 rush or in the early T2 stage.

    "Fire Beetle" Mobile Bomb (XRL0302):
        Energy consumption: 0 -> 10

Co-authored-by: Basilisk3
Co-authored-by: lL1l1

lL1l1 committed:
Fixes and QoL improvements for Megalith's build capabilities (#6559)

Make the build range longer because with 1 build range you can drag the build preview to start building at a great distance (similar to interrupt pathfinding), but without that micro the unit won't build in a sensible fashion, constantly walking around to position itself on top of or next to the build target.
Update build range preview to work off unit footprint + MaxBuildDistance + target skirt size (minimum 1, so we use 1).

Add repair capability so that it can finish eggs it started
Add reclaim capability since it can already reclaim by building eggs on top of wrecks
Add pause order since it has engineering capabilities
Fix out of date T3 mobile arty and T3 mobile AA egg cost stats.

Use AlwaysAlignToTerrain for crab eggs since they're temporary and very small, like walls, so they shouldn't flatten the terrain with FlattenSkirt.
Remove the BuildAttachBone that doesn't exist for the eggs so it will break if used.

lL1l1 committed:
Strat agility buff (#6535)

Improve the maneuverability and reliability of Strategic Bombers to promote proactive use of them and as a competing option to T3 gunships

  Strategic Bombers (UAA0304, UEA0304, URA0304, XSA0304):

  - Maneuverability:

    - Turn speed: 0.8 -> 1.2

      Lets strats micro to more effectively avoid flak damage and leaves less opportunities for ASF/SAMs to deal damage.

    - Lift Factor: 7 -> 10

      Complements the new turn speed by improving how quickly strats lift off and navigate up/down terrain.

  - Reliability:

    - Bomb drop threshold: 3.5 -> 8

      Greatly improves the reliability of bomb drops on uneven terrain or during turns. This is similar to what was done for the Ahwassa in [#2465](https://github.com/FAForever/fa/pull/2465).

      As a side effect, this usually makes the bomb drop from further away, so ACUs will have an easier time dodging bombs. Dodging was made more difficult with the strat elevation reduction in [#4799](https://github.com/FAForever/fa/pull/4799), so this counteracts that change.

    - Smaller loops on auto attack to maximize DPS:

      - `BreakOffDistance`: 60 -> 50
      - `RandomBreakOffDistanceMult`: 1.5 -> 1

      This only affects bombers automatically attacking from an idle state or from an attack, attack move, or patrol order. Previously strats would drop bombs every ~13 seconds, and now they drop every ~10 seconds.

    - Auto-attack range: 90 -> 64

      Strats will no longer automatically fly off to attack very distant targets. The 64 range matches strats' radar range.

- (#6535) Reduce the veterancy requirement for Strategic Bombers to match their mass cost reductions in previous patches.

  **Strategic Bombers (UAA0304, UEA0304, URA0304, XSA0304):**

  - Mass killed required per veterancy level: 4200 -> 2x of own mass cost (3500)

lL1l1 committed:
Give land scouts radar structure priority (#6842)

Land scouts have very low dps that is often not useful but can still kill radars, damaged units, and other land scouts. The target priorities are changed to better suit that.

Garanas committed:
Add a call to the deployment workflow of a new unit database (#6958)

Garanas committed:
Update reference to `reference`

lL1l1 committed:
Annotate command graph (#6955)

lL1l1 committed:
Fix weapons with fire rate modifiers getting stuck waiting for energy drain (#6915)

Hdt80bro committed:
Remove time requirement for rating games when recalling (#6964)

4z0t committed:
Set setons as default map for development scenarios (#6969)

BlackYps committed:
Changelog for 3829 (#6968)

Co-authored-by: Rhaelya <rhaelya@proton.me>

lL1l1 committed:
Set default tab size in markdown files to 2 (#6971)

lL1l1 committed:
Fix issues with `GiveNukeSiloBlocks` (#6855)

  - Crash if used on a unit whose first weapon does not have a `ProjectileId` in its blueprint.
  - Setting incorrect progress for units whose first weapon is not the desired silo weapon (ex: missile SACU).
  - Setting incorrect progress for units with buildrate different from their blueprint spec (ex: missile SACU).
  - Yellow work progress bar not being updated after progress is set.

maudlin27 committed:
Match ravager size category to its physical size (#6957)

4z0t committed:
Update consumption of silos when adjacent mass fabricators are toggled (#6973)

lL1l1 committed:
Fix mass fabs and extractors giving an extra adjacency buff when spawned instead of built (#6975)

"Spawned" meaning using the cheat spawn menu or transferring units between players

lL1l1 committed:
Fix `UnitDoTThread` dealing the first damage tick before waiting (#6963)

4z0t committed:
Set game speed adjustable when cheats enabled (#6978)

4z0t committed:
Annotation: `OnSelectionChanged` and `GetSelectedUnits` (#6967)

speed2CZ committed:
OpAI annotations and missing platoon templates (#6933)

lL1l1 committed:
Annotate remainder of ping.lua (#6941)

lL1l1 committed:
Fix error when ues0304 fires without a muzzle charge sequence (#6956)

speed2CZ committed:
Fix wrong army report to lobby server with uneven teams (#6948)

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

4z0t committed:
Add missing items for TechCategory (#6979)

Garanas committed:
Fix treads spawning for units that are on a transport (#6881)

Basilisk3 committed:
Increase `GuardScanRadius` limit for naval units (#6939)

Fixes certain naval units being unable to attack-move and make use of their full weapon range.

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

speed2CZ committed:
Fix ship wake effects disappearing after a while (#6981)

lL1l1 committed:
Fix unitview not showing shield HP for non-back slot shield upgrades (#6950)

Dhomie committed:
Fix missing unit param for SpecificUnitsInArea objective type (#6931)

Co-authored-by: FAForever Machine User <github@faforever.com>

Garanas committed:
Improve annotations for entity category functions (#6889)

Co-authored-by: speed2CZ <odstrcilpetr@seznam.cz>
Co-authored-by: Alexander <84857900+4z0t@users.noreply.github.com>

lL1l1 committed:
Fix modded faction icons and tooltip colors in create unit dialog (#6872)

lL1l1 committed:
Fix outdated `Health` and `DamageRadius` buffs (#6839)

lL1l1 committed:
Implement buff icon sync capability (#6838)

lL1l1 committed:
Further refinements for factory transfer (#6821)

Garanas committed:
Remove the current implementation to block paintings (#6949)

The chat configuration needs to be bugfixed so that people cannot block their own paintings and so that the mute config doesn't save across sessions.

Hdt80bro committed:
Simplify defeat backend (#6965)

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

Gatsik committed:
Fix permissions to auto-draft all PRs (#6988)

Basilisk3 committed:
Increase the `StrategicIconSize` of the Exodus's Oblivion Cannon to `3` (#6993)

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

Garanas committed:
Add lobby option to extend command sources when an army is abandoned (#6997)

Garanas committed:
Add workflow to automatically update new unit database by ETFreeman upon release (#7005)

vjt4 committed:
Reduce Sparky Guard scan radius (#7001)

Since the Sparky is an engineer it will reclaim when on patrol/attack move, however as it also has a weapon if an enemy unit gets into its guard scan radius it would stop reclaiming and instead move to engage the enemy unit. As its weapon range is only 18 but the Guard scan radius was 26 this meant that the Sparky would just being chasing units even when it was not in range. With the new Guard scan radius of 8 the Sparky should stay still to reclaim even when an enemy unit gets into its weapon range, essentially defending itself as it is reclaiming, only when an enemy gets very close it will stop reclaiming and begin to chase the enemy.

A side effect of this change is because Guard scan radius also determines the area in which an engineer reclaims when on patrol/attack move that Sparkies will walk around less and stick closer to their assigned order when looking for reclaim.

Changes:
Sparky Guard scan radius 26 -> 8

lL1l1 committed:
Reduce Deceiver stealth field range 24 -> 20 (#6972)

Reverts the deceiver range buff from the 15% vision buff (#5121 #5142) that was already reverted a while ago (#5836). This is primarily aimed at making fire beetles less powerful, but it is also a general nerf to how easily a Deceiver can cover an army.

Deceiver: T2 Mobile Stealth Field System (URL0306):

    Radar and sonar stealth field radius: 24 -> 20

lL1l1 committed:
Revert T1 interceptor and Swiftwind StartTurnDistance 10 -> 5 (#6966)

After #5042, interceptor micro feels bad due to the unintuitive nature of the huge StartTurnDistance, which is the distance within which aircraft will hover instead of turn to move to a waypoint. Most air units have a StartTurnDistance of 5 and there is no visual indicator for the distance so move orders meant to micro turns end up hovering inties, especially when large numbers of inties are in formation.

Changes:
T1 Interceptors and Swiftwind (UAA0102, UEA0102, URA0102, XSA0102, XAA0202):
StartTurnDistance: 10 -> 5

Basilisk3 committed:
Buff the Exodus's `TurretYawRange`, `TurretYawSpeed`, range, and accuracy (#6962)

This PR reverts the nerfs to the Exodus' TurretYawRange and TurretYawSpeed introduced in commit 3177522. These changes were a significant nerf to the unit's microability, giving it almost the same TurretYawSpeed as the Salem, and nearly the same TurretYawRange. It also partially reverts a previous range nerf to make it better at shore bombardment again. These changes should make microing it more rewarding again.

Exodus Class: T2 Destroyer (UAS0201):

    Oblivion Cannon
        FiringRandomness: 0.25 --> 0.2
        FiringTolerance: 2 --> 1
        MaxRadius: 70 --> 75
        TurretYawRange: 150 --> 160
        TurretYawSpeed: 90 --> 100

Basilisk3 committed:
Tweak the formula used to calculate the level of detail (LOD) for props. (#6906)

Co-authored-by: BlackYps <jonas-fb@gmx.de>

lL1l1 committed:
Fix snippets for #6997 and #7000 (#7014)

lL1l1 committed:
Revert Cybran frigate range nerf (#7019)

Reverts Cybran frigate range nerf from #4565 as it left the frigates a bit lacking in terms of being able to defend poorly-maneuverable Salems in the current destroyer-focused naval strategies. It also broke some maps which placed mexes specifically so that frigates could raid them.

Cybran Frigate URS0103
Proton Cannon MaxRadius: 26 -> 28

lL1l1 committed:
Increase fire beetle build time 550 -> 700 (#7016)

Increase fire beetle build time by reverting the previous build time decrease in #5548. Beetles have shown themselves to be extremely powerful in defensive situations and on the T2 stage. With the deceiver nerf in #6972 nerfing beetles in combat, this build time increase makes beetles require a larger infrastructure to build in large amounts for emergency defenses or for mixing into armies.

Fire Beetle: T2 Mobile Bomb (XRL0302)

    Build time: 550 -> 700

Strogoo committed:
Tread marks glitch fix (#7021)

Basilisk3 committed:
Add two missing snippets (#7023)

Strogoo committed:
Fix bombers not automatically attacking nearby enemies (#7022)

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

BlackYps committed:
Fix snippet

Rhaelya committed:
ACU TML MinRadius 30 -> 20 (#7024)

lL1l1 committed:
Make antitorps more effective against the Seraphim destroyer (#7015)

Make anti-torpedoes more effective against the Seraphim destroyer. This makes T2 subs stronger against the Seraphim destroyer, giving Aeon and Cybran navies a new counterplay option.

Uashavoh: T2 Destroyer (XSS0201)

    Ana-It Torpedo Launcher:
        Torpedoes per salvo: 4 -> 3
        Salvo delay between shots: 0.9s -> 0.8s
        Damage: 95 (95 DPS) -> 126 (94 DPS)

lL1l1 committed:
Increase Seraphim T1 sub HP 550 -> 610 (#7017)

This allows it to survive 1 torpedo bomber pass just like other T1 subs, and allows it to draw in a 1v1 against other T1 subs.

Sou-istle T1 Attack Submarine (XSS0203):

MaxHealth: 550 -> 610

Basilisk3 committed:
More snippet fixes (#7026)

Saver27 committed:
Fix Fatboy mesh assigning albedo incorrectly for sides of treads (#7008)

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

lL1l1 committed:
Re-add RackSalvoReloadTime to Seraphim MML to re-enable reload animation (#7013)

Gatsik committed:
Fallback `OptionShowPlayerNames` to 'on' when it is nil (#6987)

lL1l1 committed:
Fix profiler benchmarking sending unserializable data (#6985)

lL1l1 committed:
Fix silos with delay before firing being able to fire twice (#7010)

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

lL1l1 committed:
Add nil check to FactoryUnit OnStopBuild (#7029)

BlackYps committed:
Changelog 3830 (#7027)

lL1l1 committed:
Fix manual overcharge not working (#7031)

K-ETFreeman committed:
Fix workflow in etfreeman-db-update.yaml

Gatsik committed:
Add 'Salvation' to notify's default messages (#6959)

lL1l1 committed:
Add ability to launch custom lobby to launch script (#6856)

BlackYps committed:
Fix spooky unit db updates (#7033)

lL1l1 committed:
Fix becoming observer at end of game clearing score screen button (#7042)

Co-authored-by: (Jip) Willem Wijnia <15778155+Garanas@users.noreply.github.com>

NoryGit committed:
Display the `Minus` hotkey label correctly (#7043)

lL1l1 committed:
Fix command source not matching army index in AI games (#7038)

lL1l1 committed:
Don't show reclaim labels for unreclaimable but vulnerable props (#7034)

BlackYps committed:
Changelog 3831 (#7047)

NoryGit committed:
Salem Class: fix Nanite Torpedoes preventing other weapons from firing during their salvo (#7039)

relent0r committed:
simplify unitcat function and add some more documentation

relent0r commented:
PR is too old and the commit history is too messy now. Will create new at a later date.
